### ISSUE TITLE
Minimal notebook to apply a pretrained model

### ISSUE THREAD

**Issue by haesleinhuepf** [baigouy/EPySeg#4]
Dear Benoit @baigouy ,

first of all, congrats to EPySeg, it works great on our data! I was wondering if there is an example notebook available somewhere that is more minmal compared to [this one](https://github.com/baigouy/notebooks/blob/master/EPySeg_segment_v3.ipynb). I'm in particular interested in python code for "just" executing image segmentation / prediction to a single image. E.g. is there a way for getting a numpy array back from `deepTA.predict()` instead of it saving the result to disc? And can one create an `Img` from a numpy array? 

If possible, I would like to write a minimal notebook similar to these two for stardist and cellpose:
* https://github.com/BiAPoL/Bio-image_Analysis_with_Python/blob/main/machine_learning/stardist.ipynb
* https://github.com/BiAPoL/Bio-image_Analysis_with_Python/blob/main/machine_learning/cellpose.ipynb

Any hint is welcome!

Thanks a lot in advance and greetings from Dresden!

Best,
Robert

**Comment by baigouy** [comment]
Dear Robert @haesleinhuepf,

Thanks for using our tool. When I developped the tool, I really had in mind that users would rely on the GUI to process their data, which actually may not necessarily be the case (as I get more and more request for command line support)... Sorry, I haven't yet made any easy function function to process a single image. Actually, in my development version, I have a `predict_single` method that returns an image (ndarray as output as you want), but I haven't finalized things and even though it's functional, it's really far from optimal. Also, it's not publicly available (pushed on github) and I won't probably release it before a month. The reason for this long delay is that I'm working on a pipeline to do all the steps after segmentation, such as tracking and cell quantification and I'm not done with that yet, and I won't do a release before that is ready too. In the meantime, I could provide you with a' quick and very dirty' minimal code to produce single image segmentation, if you wish, but that would definitely not look nice and simple in a notebook, especially compared to the exemples you provide (because I need define new functions,...).
In any case, I will definitely keep your request in mind and I'll add a very simple way to call the epyseg segmentation in the the future and I'll keep you posted when it's done.
Regarding the Img object, indeed, one can create an `Img` from a numpy array, e.g.:
`img = Img(ndarray, dimensions='hwc')`


All the best,

Benoit

ps: say hi for me to all the DD guys

**Comment by haesleinhuepf** [comment]
Hi Benoit @baigouy ,

> In the meantime, I could provide you with a' quick and very dirty' minimal code to produce single image segmentation, if you wish, but that would definitely not look nice and simple in a notebook, especially compared to the exemples you provide (because I need define new functions,...).

I would love to have that! I can also wrap this function into a function and then hide it from a notebook. You know we're mostly interested in "just" applying multiple segmentation algorithms and then pick the best for a given project. And there are plenty of projects around, as you know I guess. ;-)

Thanks!
Robert

**Comment by baigouy** [comment]
Hi Robert,

Here you go (see __main__), that should work (but it's definitely very
poorly made and suboptimal):

<snip>


import traceback

import matplotlib.pyplot as plt

from epyseg.deeplearning.callbacks.stop import myStopCallback
from epyseg.deeplearning.deepl import EZDeepLearning
import os
import numpy as np

from epyseg.img import Img
from epyseg.postprocess.filtermask import simpleFilter

def predict_single_image(deepTA,
single_image_predict_generator,predict_parameters, batch_size=1):
                         # progress_callback=None,
                   # hq_predictions='mean',
post_process_algorithm=None, hq_pred_options='all',
                   # **kwargs):

    input_shape = deepTA.get_inputs_shape()
    output_shape = deepTA.get_outputs_shape()

    predict_generator = deepTA.get_predict_generator(
        inputs=[single_image_predict_generator], input_shape=input_shape,
        output_shape=output_shape,
        z_frames_to_add=None,
        clip_by_frequency=None, **predict_parameters)  # before was
0.05 which is bad especially with restored images

    results = None
    image, crop_parameters = next(predict_generator.predict_generator())
    try:
        results = deepTA.model.predict(image, verbose=1, batch_size=batch_size)
        if predict_parameters["hq_pred_options"] is not None:
            # TODO fix the line below to handle several outputs
support --> TODO (do that in a smart way)
            results = deepTA.get_HQ_predictions(image, results,
batch_size=batch_size,

projection_method=predict_parameters["hq_pred_options"],
                                              hq_pred_options=
predict_parameters["hq_pred_options"] )
    except:
        traceback.print_exc()
        print('Could not predict output for cur image')

    if results is None:
        print('Prediction interrupted or failed. Stopping...')
        return

    if isinstance(results, np.ndarray):  # convert to list if single output
        results = [results]


    # need rempve the loops in single since only one image and one
data --> redo the code
    final_output = []
    for cur_count, result in enumerate(results):
        for j in range(len(crop_parameters)):
            # cur_output_shape = output_shape[cur_count]
            ordered_tiles = Img.linear_to_2D_tiles(result, crop_parameters[j])

            # 2D image
            # if len(cur_output_shape) == 4:
            reconstructed_tile = Img.reassemble_tiles(ordered_tiles,
crop_parameters[j])
            # run post process directly on the image if available
            # if cur_output_shape[-1] != 7 and
(predict_parameters["post_process_algorithm"] is not None and (
            #
isinstance(predict_parameters["post_process_algorithm"], str) and not
(
            #         'imply' in
predict_parameters["post_process_algorithm"] or 'first' in
predict_parameters["post_process_algorithm"]))):
            #     print('Model is not compatible with epyseg and
cannot be optimized, so the desired post processing cannot be applied,
sorry...')
            #
            # if
isinstance(predict_parameters["post_process_algorithm"],str) and
'imply' in predict_parameters["post_process_algorithm"]:
            #     reconstructed_tile =
simpleFilter(Img(reconstructed_tile, dimensions='hwc'),
**predict_parameters)
            #     final_output.append(reconstructed_tile)
            # elif predict_parameters["post_process_algorithm"] is not None:
            try:
                print('post processing/refining mask, please wait...')
                reconstructed_tile =
deepTA.run_post_process(Img(reconstructed_tile, dimensions='hwc'),
                                                           #
predict_parameters["post_process_algorithm"],

progress_callback=None,
                                                           **predict_parameters)
                final_output.append(reconstructed_tile)
            except:
                print('running post processing/refine mask failed')
                traceback.print_exc()
                # else:
                #     final_output.append(reconstructed_tile)
            # else:
            #     # 3D image
            #     reconstructed_tile =
Img.reassemble_tiles(ordered_tiles, crop_parameters[j], three_d=True)
            #     # run post process directly on the image if available
            #     if cur_output_shape[-1] != 7 and
(predict_parameters["post_process_algorithm"] is not None or (
            #
isinstance(predict_parameters["post_process_algorithm"], str) and
'imply' in predict_parameters["post_process_algorithm"])):
            #         print(
            #             'Model is not compatible with epyseg and
cannot be optimized, so it will simply be thresholded according to
selected options, sorry...')
            #     if isinstance(predict_parameters["post_process_algorithm"],
            #                   str) and 'imply' in
predict_parameters["post_process_algorithm"]:
            #         print('not supported yet please threshold
outside the software')
            #         final_output.append(reconstructed_tile)
            #     elif predict_parameters["post_process_algorithm"] is not None:
            #         try:
            #             print('post processing/refining mask, please wait...')
            #             reconstructed_tile =
deepTA.run_post_process(Img(reconstructed_tile, dimensions='dhwc'),
            #
predict_parameters["post_process_algorithm"],
            #
progress_callback=predict_parameters["post_process_algorithm"],
            #
**predict_parameters)
            #             final_output.append(reconstructed_tile)
            #         except:
            #             print('running post processing/refine mask failed')
            #             traceback.print_exc()
            #     else:
            #         final_output.append(reconstructed_tile)

    return final_output[0] # very dirty and loop is useless in this case

if __name__ == '__main__':
    INPUT_FILE = ['/E/trash/test2.png', '/E/trash/test.tif']
    # INPUT_FILE = '/path/to/files_to_segment/'
    input_channel_of_interest = None  # assumes image is single
channel or multichannel nut channel of interest is ch0, needs be
changed otherwise, e.g. 1 for channel 1
    TILE_WIDTH = 256  # 128 # 64
    TILE_HEIGHT = 256  # 128 # 64
    TILE_OVERLAP = 32
    EPYSEG_PRETRAINING = 'Linknet-vgg16-sigmoid-v2'  # or
'Linknet-vgg16-sigmoid' for v1
    SIZE_FILTER = None  # 100 # set to 100 to get rid of cells having
pixel area < 100 pixels

    # do I have a bug in my tiling algo it does not seem to give the
right image size...

    predict_parameters = {}
    input_normalization = {'method': 'Rescaling (min-max
normalization)', 'range': [0, 1], 'individual_channels': True}
    predict_parameters["input_channel_of_interest"] = input_channel_of_interest
    predict_parameters["default_input_tile_width"] = TILE_WIDTH
    predict_parameters["default_input_tile_height"] = TILE_HEIGHT
    predict_parameters["default_output_tile_width"] = TILE_WIDTH
    predict_parameters["default_output_tile_height"] = TILE_HEIGHT
    predict_parameters["tile_width_overlap"] = TILE_OVERLAP
    predict_parameters["tile_height_overlap"] = TILE_OVERLAP
    predict_parameters["hq_pred_options"] = "Use all augs (pixel
preserving + deteriorating) (Recommended for segmentation)"  # not yet
supported on current versions of epyseg
    predict_parameters["post_process_algorithm"] = "default
(slow/robust) (epyseg pre-trained model only!)"  # None
    predict_parameters["input_normalization"] = input_normalization
    predict_parameters["filter"] = SIZE_FILTER
    predict_parameters["hq_predictions"] = 'mean'

    deepTA = EZDeepLearning()
    deepTA.load_or_build(architecture='Linknet', backbone='vgg16',
activation='sigmoid', classes=7, pretraining=EPYSEG_PRETRAINING)
    # deepTA.get_loaded_model_params()
    deepTA.stop_cbk = myStopCallback()

    for file in INPUT_FILE:
        # data = run_seg(deepTA=deepTA, INPUT_FOLDER=[file],
input_channel_of_interest=input_channel_of_interest,
TILE_WIDTH=TILE_WIDTH, TILE_HEIGHT=TILE_HEIGHT,
TILE_OVERLAP=TILE_OVERLAP,EPYSEG_PRETRAINING=EPYSEG_PRETRAINING,
SIZE_FILTER=SIZE_FILTER )
        data = predict_single_image(deepTA, file,
predict_parameters=predict_parameters, batch_size=1)
        plt.imshow(data)
        plt.show()


<snap>

all the best,

Benoit


Le mer. 27 oct. 2021 à 14:19, Robert Haase ***@***.***> a
écrit :

**Comment by alvillars** [comment]
Hi Benoit @baigouy and Robert @haesleinhuepf I was interesting in something similar. Thank you for sharing. May I use this as well? Regarding your comment @haesleinhuepf would you be willing to share such notebook comparing different segmentation techniques? All the best, Alexis.

**Comment by baigouy** [comment]
Yes, please do!

Le jeu. 28 oct. 2021 à 08:56, Alexis Villars ***@***.***> a
écrit :

**Comment by Parsa744** [comment]
Hi dear Benoit @baigouy,
When I want to run your minimal code I always get  'ERROR - 2023-02-06 17:24:47,144 - deepl.py - load_weights - line 847 - Error --> try loading weights by name, i.e. model and weights don't match ???' 
and 
'ERROR - 2023-02-06 17:24:59,844 - deepl.py - load_weights - line 850 - Definitely failed loading weights, there is no match between weights and model!!!
'
this error. Do you have any idea why?

**Comment by baigouy** [comment]
Dear @Parsa744,

Thanks for your message. I could run it on colab:
https://github.com/baigouy/notebooks/blob/master/old/minimal.ipynb
I just had to change the path in the 'INPUT_FILE = ...' variable at the end of the file

hope this helps

Best,

Benoit
